plugins/env/app/models/environment_variable.rb

```diff
@@ -1,16 +1,41 @@
 class EnvironmentVariable < ActiveRecord::Base
   belongs_to :parent, polymorphic: true
-  belongs_to :deploy_group
+  belongs_to :scope, polymorphic: true
   validates :name, presence: true
+  validates :scope_type, inclusion: ["Environment", "DeployGroup", nil]
 
-  def self.env(stage, deploy_group_id)
+  def self.env(stage, deploy_group)
     variables = stage.environment_variables + stage.environment_variable_groups.flat_map(&:environment_variables)
+    variables.sort_by! { |ev| [ev.parent_type == "Stage" ? 1 : 0, ev.send(:priority)] }
     variables.each_with_object({}) do |ev, all|
-      if !ev.deploy_group_id
-        all[ev.name] ||= ev.value
-      elsif ev.deploy_group_id == deploy_group_id
-        all[ev.name] = ev.value
-      end
+      match = (
+        !ev.scope_id || # for all
+        (deploy_group && (
+          (ev.scope_type == "DeployGroup" && ev.scope_id == deploy_group.id) || # matches deploy group
+          (ev.scope_type == "Environment" && ev.scope_id == deploy_group.environment_id) # matches deploy group's environment
+        ))
+      )
+      all[ev.name] = ev.value if match
+    end
+  end
+
+  # used to assign direct from form values
+  def scope_type_and_id=(value)
```

Comment:
this kinda seems like controller logic since its deserializing POST / PUT requests